Microsoft.Azure.Cosmos/src/Serializer/CosmosSystemTextJsonSerializer.cs

```
@@ -49,8 +52,28 @@ public override T FromStream<T>(Stream stream)
 
             using (stream)
             {
-                using StreamReader reader = new (stream);
-                return JsonSerializer.Deserialize<T>(reader.ReadToEnd(), this.jsonSerializerOptions);
+                if (stream is Documents.CloneableStream cloneableStream)
+                {
+                    using (CosmosBufferedStreamWrapper bufferedStream = new (cloneableStream, shouldDisposeInnerStream: false))
+                    {
+                        if (bufferedStream.GetJsonSerializationFormat() == JsonSerializationFormat.Binary)
+                        {
+                            byte[] content = bufferedStream.ReadAll();
+
+                            if (CosmosObject.TryCreateFromBuffer(content, out CosmosObject cosmosObject))
+                            {
+                                return System.Text.Json.JsonSerializer.Deserialize<T>(cosmosObject.ToString(), this.jsonSerializerOptions);
+                            }
+                            else
+                            {
+                                using Stream textStream = CosmosSerializationUtil.ConvertToStreamUsingJsonSerializationFormat(content, JsonSerializationFormat.Text);
+                                return this.DeserializeStream<T>(textStream);
```

What would be the case for TryCreate to fail? 
I am wondering if we can safely error out as unexpected?

---

Because we had some of the known serialization gaps, I kept this `else` condition to safe guard us to at-least use another mechanism to deserialize. Fixing the serialization gaps are tracked as a part of this [task](https://github.com/Azure/azure-cosmos-dotnet-v3/issues/4719).

Ideally, if there are any legitimate issue, then both the conditions will fail. The else condition is there to cover any known gaps, until we fix them as a part of the above task.